{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(Biostrings)\n",
    "library(lessR)\n",
    "library(vegan)\n",
    "library(phyloseq)\n",
    "library(ggplot2)\n",
    "library(ape)\n",
    "library(picante)\n",
    "library(ggpubr)\n",
    "library(readxl)\n",
    "library(devtools)\n",
    "library(pairwiseAdonis)\n",
    "library(compositions)\n",
    "library(microbiome)\n",
    "library(readxl)\n",
    "library(dbplyr)\n",
    "library(dplyr)\n",
    "\n",
    "\n",
    "setwd(\"~/Documents/analysis/\")\n",
    "path <- \"~/Documents/analysis/\"\n",
    "\n",
    "\n",
    "#Import data into phyloseq\n",
    "tree <- read.nexus(file=\"2038.tre\")\n",
    "class(tree)\n",
    "\n",
    "#import taxa\n",
    "all_taxa_table<- as.matrix(read.csv(\"hannataxa.csv\", \n",
    "                                    sep=\",\", header=T, row.names=1))\n",
    "dim(all_taxa_table)\n",
    "otu <- as.matrix(read.csv(\"hanna_counts.csv\", \n",
    "                           sep=\",\", header=T, row.names=1))\n",
    "dim(otu)\n",
    "\n",
    "All_seqs <- readDNAStringSet(\"All_tab_nochim.fa\",format = \"fasta\")\n",
    "length(All_seqs)\n",
    "\n",
    "OTU = otu_table(otu, taxa_are_rows = TRUE)\n",
    "dim(OTU)\n",
    "\n",
    "metadata11 <- read_excel(\"metadata11.xlsx\")\n",
    "#View(metadata11)\n",
    "meta <- as.data.frame(metadata11)\n",
    "rownames(meta) <- meta$GroupID\n",
    "dim(meta)\n",
    "#View(meta)\n",
    "\n",
    "All_samples_phylo <- phyloseq(otu_table(OTU), \n",
    "                              tax_table(all_taxa_table), \n",
    "                              phy_tree(tree), refseq(All_seqs), \n",
    "                              sample_data(meta))\n",
    "All_samples_phylo\n",
    "All_samples_phylo1 <- subset_taxa(All_samples_phylo, !is.na(Phylum) & !Phylum %in% c(\"\", \"uncharacterized\"))\n",
    "All_samples_phylo1\n",
    "#1898 ASVs\n",
    "All_samples_phylo1@sam_data$julian <- as.numeric(as.character(All_samples_phylo1@sam_data$julian))\n",
    "All_samples_phylo1@sam_data$Reduced <- as.numeric(as.character(All_samples_phylo1@sam_data$Reduced))\n",
    "All_samples_phylo1@sam_data$pairedReads <- as.numeric(as.character(All_samples_phylo1@sam_data$pairedReads))\n",
    "\n",
    "TAXe = as(tax_table(All_samples_phylo1), \"matrix\")\n",
    "write.csv(TAXe, \"correct_tax.csv\" )\n",
    "\n",
    "# Extract abundance matrix from the phyloseq object\n",
    "OTU1 = as(otu_table(All_samples_phylo1), \"matrix\")\n",
    "# transpose if necessary\n",
    "if(taxa_are_rows(All_samples_phylo1)){OTU1 <- t(OTU1)}\n",
    "# Coerce to data.frame\n",
    "OTUdf = as.data.frame(OTU1)\n",
    "View(OTUdf)\n",
    "dim(OTUdf)\n",
    "write.csv(file = \"OTU_reduced_all.csv\", t(OTUdf))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### pruning samples with low counts\n",
    "pruned_samples <- c(\"hdmarch_2018_339\",\n",
    "                    \"hdmarch_2018_343\",\n",
    "                    \"hdmarch_2018_98\",\n",
    "                    \"hdmarch_2018_86\",\n",
    "                    \"hdmarch_2018_84\",\n",
    "                    \"hdmarch_2018_5\",\n",
    "                    \"hdmarch_2018_46\",\n",
    "                    \"hdmarch_2018_35\",\n",
    "                    \"hdmarch_2018_59\",\n",
    "                    \"hdmarch_2018_48\",\n",
    "                    \"hdmarch_2018_25\",\n",
    "                    \"hdmarch_2018_224\",\n",
    "                    \"hdmarch_2018_263\",\n",
    "                    \"hdmarch_2018_81\",\n",
    "                    \"hdmarch_2018_369\",\n",
    "                    \"hdmarch_2018_259\",\n",
    "                    \"hdmarch_2018_82\",\n",
    "                    \"hdmarch_2018_307\",\n",
    "                    \"hdmarch_2018_83\",\n",
    "                    \"hdmarch_2018_336\",\n",
    "                    \"hdmarch_2018_261\",\n",
    "                    \"hdmarch_2018_234\",\n",
    "                    \"hdmarch_2018_296\",\n",
    "                    \"hdmarch_2018_229\",\n",
    "                    \"hdmarch_2018_214\",\n",
    "                    \"hdmarch_2018_309\",\n",
    "                    \"hdmarch_2018_329\",\n",
    "                    \"hdmarch_2018_80\",\n",
    "                    \"hdmarch_2018_274\",\n",
    "                    \"hdmarch_2018_249\",\n",
    "                    \"hdmarch_2018_271\",\n",
    "                    \"hdmarch_2018_273\",\n",
    "                    \"hdmarch_2018_250\",\n",
    "                    \"hdmarch_2018_294\",\n",
    "                    \"hdmarch_2018_228\",\n",
    "                    \"hdmarch_2018_254\",\n",
    "                    \"hdmarch_2018_210\",\n",
    "                    \"hdmarch_2018_90\",\n",
    "                    \"hdmarch_2018_376\",\n",
    "                    \"hdmarch_2018_295\",\n",
    "                    \"hdmarch_2018_278\",\n",
    "                    \"hdmarch_2018_332\",\n",
    "                    \"hdmarch_2018_297\",\n",
    "                    \"hdmarch_2018_272\",\n",
    "                    \"hdmarch_2018_258\",\n",
    "                    \"hdmarch_2018_298\",\n",
    "                    \"hdmarch_2018_360\",\n",
    "                    \"hdmarch_2018_262\",\n",
    "                    \"hdmarch_2018_238\",\n",
    "                    \"hdmarch_2018_264\",\n",
    "                    \"hdmarch_2018_240\",\n",
    "                    \"hdmarch_2018_78\",\n",
    "                    \"hdmarch_2018_76\",\n",
    "                    \"hdmarch_2018_222\",\n",
    "                    \"hdmarch_2018_79\",\n",
    "                    \"hdmarch_2018_204\",\n",
    "                    \"hdmarch_2018_227\",\n",
    "                    \"hdmarch_2018_231\",\n",
    "                    \"hdmarch_2018_72\",\n",
    "                    \"hdmarch_2018_75\",\n",
    "                    \"hdmarch_2018_71\",\n",
    "                    \"hdmarch_2018_36\",\n",
    "                    \"hdmarch_2018_51\",\n",
    "                    \"hdmarch_2018_50\",\n",
    "                    \"hdmarch_2018_66\",\n",
    "                    \"hdmarch_2018_40\",\n",
    "                    \"hdmarch_2018_22\",\n",
    "                    \"hdmarch_2018_30\",\n",
    "                    \"hdmarch_2018_91\",\n",
    "                    \"hdmarch_2018_26\",\n",
    "                    \"hdmarch_2018_4\",\n",
    "                    \"hdmarch_2018_24\",\n",
    "                    \"hdmarch_2018_61\",\n",
    "                    \"hdmarch_2018_64\",\n",
    "                    \"hdmarch_2018_23\",\n",
    "                    \"hdmarch_2018_7\",\n",
    "                    \"hdmarch_2018_3\",\n",
    "                    \"hdmarch_2018_223\",\n",
    "                    \"hdmarch_2018_34\",\n",
    "                    \"hdmarch_2018_47\",\n",
    "                    \"hdmarch_2018_27\",\n",
    "                    \"hdmarch_2018_28\",\n",
    "                    \"hdmarch_2018_275\",\n",
    "                    \"hdmarch_2018_33\",\n",
    "                    \"hdmarch_2018_52\",\n",
    "                    \"hdmarch_2018_280\",\n",
    "                    \"hdmarch_2018_38\",\n",
    "                    \"hdmarch_2018_29\",\n",
    "                    \"hdmarch_2018_44\",\n",
    "                    \"hdmarch_2018_60\",\n",
    "                    \"hdmarch_2018_221\",\n",
    "                    \"hdmarch_2018_370\",\n",
    "                    \"hdmarch_2018_225\",\n",
    "                    \"hdmarch_2018_208\",\n",
    "                    \"hdmarch_2018_251\",\n",
    "                    \"hdmarch_2018_203\",\n",
    "                    \"hdmarch_2018_85\",\n",
    "                    \"hdmarch_2018_119\",\n",
    "                    \"hdmarch_2018_186\",\n",
    "                    \"hdmarch_2018_151\",\n",
    "                    \"hdmarch_2018_121\",\n",
    "                    \"hdmarch_2018_143\",\n",
    "                    \"hdmarch_2018_19\",\n",
    "                    \"hdmarch_2018_200\",\n",
    "                    \"hdmarch_2018_109\",\n",
    "                    \"hdmarch_2018_116\",\n",
    "                    \"hdmarch_2018_155\",\n",
    "                    \"hdmarch_2018_164\",\n",
    "                    \"hdmarch_2018_190\",\n",
    "                    \"hdmarch_2018_146\",\n",
    "                    \"hdmarch_2018_167\",\n",
    "                    \"hdmarch_2018_108\",\n",
    "                    \"hdmarch_2018_189\",\n",
    "                    \"hdmarch_2018_154\",\n",
    "                    \"hdmarch_2018_127\",\n",
    "                    \"hdmarch_2018_145\",\n",
    "                    \"hdmarch_2018_191\",\n",
    "                    \"hdmarch_2018_141\",\n",
    "                    \"hdmarch_2018_106\",\n",
    "                    \"hdmarch_2018_152\",\n",
    "                    \"hdmarch_2018_149\",\n",
    "                    \"hdmarch_2018_114\",\n",
    "                    \"hdmarch_2018_137\",\n",
    "                    \"hdmarch_2018_201\",\n",
    "                    \"hdmarch_2018_182\",\n",
    "                    \"hdmarch_2018_138\",\n",
    "                    \"hdmarch_2018_150\",\n",
    "                    \"hdmarch_2018_110\",\n",
    "                    \"hdmarch_2018_179\",\n",
    "                    \"hdmarch_2018_126\",\n",
    "                    \"hdmarch_2018_104\",\n",
    "                    \"hdmarch_2018_177\",\n",
    "                    \"hdmarch_2018_100\",\n",
    "                    \"hdmarch_2018_133\",\n",
    "                    \"hdmarch_2018_101\",\n",
    "                    \"hdmarch_2018_148\",\n",
    "                    \"hdmarch_2018_130\",\n",
    "                    \"hdmarch_2018_123\",\n",
    "                    \"hdmarch_2018_125\",\n",
    "                    \"hdmarch_2018_1\",\n",
    "                    \"hdmarch_2018_16\",\n",
    "                    \"hdmarch_2018_12\",\n",
    "                    \"hdmarch_2018_17\",\n",
    "                    \"hdmarch_2018_11\",\n",
    "                    \"hdmarch_2018_20\",\n",
    "                    \"hdmarch_2018_2\",\n",
    "                    \"hdmarch_2018_18\",\n",
    "                    \"hdmarch_2018_10\",\n",
    "                    \"hdmarch_2018_124\",\n",
    "                    \"hdmarch_2018_142\")\n",
    "length(pruned_samples)\n",
    "pruned_samples_phylo <- prune_samples(pruned_samples, All_samples_phylo1)\n",
    "pruned_samples_phylo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get otu df from pruned samples phylo\n",
    "OTUe = as(otu_table(pruned_samples_phylo), \"matrix\")\n",
    "# transpose if necessary\n",
    "if(taxa_are_rows(pruned_samples_phylo)){OTUe <- t(OTUe)}\n",
    "# Coerce to data.frame\n",
    "OTUee = as.data.frame(OTUe)\n",
    "#View(OTUee)\n",
    "dim(OTUee)\n",
    "#View(pruned_samples_phylo@tax_table)\n",
    "TAXee = as(tax_table(pruned_samples_phylo), \"matrix\")\n",
    "#View(TAXee)\n",
    "#remove rare taxa\n",
    "pruned_samples_phylo1 <- prune_taxa(taxa_sums(pruned_samples_phylo) > 1, pruned_samples_phylo)\n",
    "pruned_samples_phylo1\n",
    "\n",
    "#get metadata table\n",
    "length(sample_names(pruned_samples_phylo1))\n",
    "pruned_samples_meta <- as.matrix(pruned_samples_phylo1@sam_data, stringsAsFactors=T)\n",
    "pruned_samples_meta <- as.data.frame(pruned_samples_meta, stringsAsFactors=T)\n",
    "pruned_samples_meta$pairedReads <- as.numeric(as.character(pruned_samples_meta$pairedReads))\n",
    "pruned_samples_meta$Reduced <- as.numeric(as.character(pruned_samples_meta$Reduced))\n",
    "pruned_samples_meta$julian <- as.numeric(as.character(pruned_samples_meta$julian))\n",
    "dim(pruned_samples_meta)\n",
    "#View(pruned_samples_meta)\n",
    "sapply(pruned_samples_meta, class)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#transform\n",
    "clr_phylo <- transform(pruned_samples_phylo1, transform = \"clr\", target = \"sample\")\n",
    "clr_phylo\n",
    "clr_phylo_otu <- as.data.frame(clr_phylo@otu_table)\n",
    "clr_phylo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get distance\n",
    "t_clr_phylo_otu <- t(clr_phylo_otu)\n",
    "clr_euc_all <- vegdist(t_clr_phylo_otu, method = \"euclidean\")\n",
    "clr_euc_all\n",
    "clr_phylo_meta <- as.matrix(clr_phylo@sam_data, stringsAsFactors=T)\n",
    "clr_phylo_meta <- as.data.frame(clr_phylo_meta, stringsAsFactors=T)\n",
    "clr_phylo_meta$pairedReads <- as.numeric(as.character(clr_phylo_meta$pairedReads))\n",
    "clr_phylo_meta$Reduced <- as.numeric(as.character(clr_phylo_meta$Reduced))\n",
    "clr_phylo_meta$julian <- as.numeric(as.character(clr_phylo_meta$julian))\n",
    "dim(clr_phylo_meta)\n",
    "View(clr_phylo_meta)\n",
    "sapply(clr_phylo_meta, class)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tests\n",
    "v <- betadisper(clr_euc_all, clr_phylo_meta$sample_type, type = c(\"median\",\"centroid\"))\n",
    "permutest(v)\n",
    "#yes\n",
    "adonis2(clr_euc_all ~ pairedReads, data = clr_phylo_meta)\n",
    "#sig\n",
    "adonis2(clr_euc_all ~ Reduced, data = clr_phylo_meta)\n",
    "#sig\n",
    "adonis2(clr_euc_all ~ sample_type, data = clr_phylo_meta)\n",
    "adonis2(clr_euc_all ~ fraction, data = clr_phylo_meta)\n",
    "adonis2(clr_euc_all ~ sample_type * pairedReads, data = clr_phylo_meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#reads by group and compare\n",
    "clr_phylo_meta2$group <- ordered(clr_phylo_meta2$sample_type,\n",
    "                         levels = c(\"Algae\", \"Raw Intake\", \"Treated Intake\", \"Waste\", \"Larvae Tank\"))\n",
    "\n",
    "group_by(clr_phylo_meta2, group) %>%\n",
    "  summarise(\n",
    "    count = n(),\n",
    "    mean = mean(pairedReads, na.rm = TRUE),\n",
    "    sd = sd(pairedReads, na.rm = TRUE),\n",
    "    median = median(pairedReads, na.rm = TRUE),\n",
    "    IQR = IQR(pairedReads, na.rm = TRUE)\n",
    "  )\n",
    "\n",
    "View(meta)\n",
    "?aov\n",
    "t <- aov(data = meta, Reduced ~ sample_type)\n",
    "class(t)\n",
    "summary(t)\n",
    "\n",
    "library(\"ggpubr\")\n",
    "ggboxplot(clr_phylo_meta2, x = \"group\", y = \"Reduced\", \n",
    "          color = \"group\",\n",
    "          order = c(\"Algae\", \"Raw Intake\", \"Treated Intake\", \"Waste\", \"Larvae Tank\"),\n",
    "          ylab = \"Reduced\", xlab = \"sample type\")\n",
    "kruskal.test(Reduced ~ group, data = clr_phylo_meta2)\n",
    "#not sig\n",
    "#sig for paired reads \n",
    "pairwise.wilcox.test(clr_phylo_meta2$pairedReads, clr_phylo_meta2$group,\n",
    "                     p.adjust.method = \"BH\")\n",
    "#sig difference in intake vs algae and treated vs larvae tank "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ordinate (entire group)\n",
    "#PCA via phyloseq\n",
    "OK <- ordinate(clr_phylo, \"RDA\", \"euclidean\")\n",
    "OK\n",
    "#Plot scree plot\n",
    "plot_scree(OK) + geom_bar(stat=\"identity\", fill = \"blue\") +\n",
    "  labs(x = \"\\nAxis\", y = \"Proportion of Variance\\n\")\n",
    "#Examine eigenvalues and % prop. variance explained\n",
    "head(OK$CA$eig)   \n",
    "sapply(OK$CA$eig[1:5], function(x) x / sum(OK$CA$eig)) \n",
    "#Scale axes and plot ordination\n",
    "clr1 <- OK$CA$eig[1] / sum(OK$CA$eig)\n",
    "clr2 <- OK$CA$eig[2] / sum(OK$CA$eig)\n",
    "\n",
    "?scores\n",
    "pl <- as.data.frame(scores(OK, display = \"sites\"))\n",
    "pll <- as.data.frame(pl)\n",
    "View(pll)\n",
    "\n",
    "hanPalette1 <- c(\"chartreuse4\", \"goldenrod1\", \"royalblue\", \"lightskyblue\",  \"thistle4\")\n",
    "\n",
    "\n",
    "p2 <- plot_ordination(clr_phylo, OK, \"samples\", color=\"sample_type\", shape = \"fraction\", axes = c(1, 2)) +\n",
    "  coord_fixed(clr2 / clr1) +\n",
    "  geom_point(color = \"black\", size=2.6) +  \n",
    "  geom_point(size = 2, alpha = 0.95) +\n",
    "  scale_color_manual(values = hanPalette1) +\n",
    "  theme_bw() + labs(fill = \"Day\") +\n",
    "  labs(title=\"Aquaculture Facility, all samples\") + \n",
    "  theme(legend.text =element_text(size=6)) +\n",
    "  theme(legend.title = element_text(size=6)) +\n",
    "  theme(axis.title.x = element_text(size = 8)) +\n",
    "  theme(axis.title.y = element_text(size = 8)) +\n",
    "  theme(axis.text.x = element_text(size =6)) +\n",
    "  theme(axis.text.y = element_text(size =6)) + theme(plot.title = element_text(hjust = 0.5)) \n",
    "p2 \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Free-living - All\n",
    "fl_phy <- pruned_samples_phylo1 %>% \n",
    "  phyloseq::subset_samples(fraction %in% \"Supor 200\") \n",
    "fl_phy2 <- fl_phy %>% \n",
    "  phyloseq::subset_samples(sample_type %in% c(\"Larvae Tank\", \"Treated Intake\", \"Raw Intake\", \"Algae\", 'Waste')) \n",
    "fl_phy2\n",
    "#get ASV table \n",
    "OTUe = as(otu_table(fl_phy2), \"matrix\")\n",
    "# transpose if necessary\n",
    "if(taxa_are_rows(fl_phy2)){OTUe <- t(OTUe)}\n",
    "# Coerce to data.frame\n",
    "OTUee = as.data.frame(OTUe)\n",
    "View(OTUee)\n",
    "dim(OTUee)\n",
    "TAXee = as(tax_table(fl_phy2), \"matrix\")\n",
    "#Revmove ASVs with no presence\n",
    "fl_phy3 <- prune_taxa(taxa_sums(fl_phy2) > 1, fl_phy2)\n",
    "#transform\n",
    "fl_clr <- transform(fl_phy3, transform = \"clr\", target = \"sample\")\n",
    "fl_phylo_otu <- as.data.frame(fl_phylo@otu_table)\n",
    "dim(fl_phylo_otu)\n",
    "View(fl_phylo_otu)\n",
    "#get metadata table\n",
    "meta <- as.matrix(fl_phylo@sam_data, stringsAsFactors=T)\n",
    "meta <- as.data.frame(meta, stringsAsFactors=T)\n",
    "meta$pairedReads <- as.numeric(as.character(meta$pairedReads))\n",
    "meta$Reduced <- as.numeric(as.character(meta$Reduced))\n",
    "meta$julian <- as.numeric(as.character(meta$julian))\n",
    "dim(meta)\n",
    "View(meta)\n",
    "#View(pruned_samples_meta)\n",
    "sapply(meta, class)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##What if I subsetted the CLR transformed phyloseq object\n",
    "clr_phylo_pa <- clr_phylo %>% \n",
    "  phyloseq::subset_samples(fraction %in% \"GFF\") \n",
    "clr_phylo_pa\n",
    "clr_phylo_pa_a <- clr_phylo_pa %>% \n",
    "  phyloseq::subset_samples(Type2 %in% c(\"Tank\", \"Intake\")) \n",
    "clr_phylo_pa_a\n",
    "#get ASV table \n",
    "OTUe = as(otu_table(clr_phylo_pa_a), \"matrix\")\n",
    "# transpose if necessary\n",
    "if(taxa_are_rows(clr_phylo_pa_a)){OTUe <- t(OTUe)}\n",
    "# Coerce to data.frame\n",
    "OTUee = as.data.frame(OTUe)\n",
    "View(OTUee)\n",
    "dim(OTUee)\n",
    "TAXee = as(tax_table(clr_phylo_pa_a), \"matrix\")\n",
    "#Revmove ASVs with no presence\n",
    "clr_phylo_pa_b <- prune_taxa(taxa_sums(clr_phylo_pa_a) > 1, clr_phylo_pa_a)\n",
    "\n",
    "#get metadata table\n",
    "meta <- as.matrix(clr_phylo_pa_b@sam_data, stringsAsFactors=T)\n",
    "meta <- as.data.frame(meta, stringsAsFactors=T)\n",
    "meta$pairedReads <- as.numeric(as.character(meta$pairedReads))\n",
    "meta$Reduced <- as.numeric(as.character(meta$Reduced))\n",
    "meta$julian <- as.numeric(as.character(meta$julian))\n",
    "dim(meta)\n",
    "View(meta)\n",
    "#View(pruned_samples_meta)\n",
    "sapply(meta, class)\n",
    "\n",
    "#distance\n",
    "clr_phylo_pa_b_otu <- as.data.frame(clr_phylo_pa_b@otu_table)\n",
    "clr_phylo_pa_b_otu_t <- t(clr_phylo_pa_b_otu)\n",
    "euc_pa <- vegdist(clr_phylo_pa_b_otu_t, method = \"euclidean\")\n",
    "#tests\n",
    "\n",
    "\n",
    "adonis2(euc_pa ~ sample_type, data = meta)\n",
    "adonis2(euc_pa ~ Type2, data = meta)\n",
    "adonis2(euc_pa ~ Type, data = meta)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
